### PR TITLE
fix: Add gas metering to BlockBeforeSend

### PR DESCRIPTION
**Pull request by mattverse**
Closes: #XXX

## What is the purpose of the change

Adds gas metering to BlockBeforeSend.

## Testing and Verifying

Add relevant test cases. 

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

**Comment by mattverse** [comment]
devbot help

File: x/tokenfactory/keeper/before_send.go [hunk 1]
```diff
@@ -133,22 +133,14 @@ func (k Keeper) callBeforeSendListener(ctx sdk.Context, from, to sdk.AccAddress,
...
-			} else {
-				childCtx := ctx.WithGasMeter(sdk.NewGasMeter(types.TrackBeforeSendGasLimit))
-				_, err = k.contractKeeper.Sudo(childCtx.WithEventManager(em), cwAddr, msgBz)
-				if err != nil {
-					return errorsmod.Wrapf(err, "failed to call before send hook for denom %s", coin.Denom)
-				}
-
-				// consume gas used for calling contract to the parent ctx
-				ctx.GasMeter().ConsumeGas(childCtx.GasMeter().GasConsumed(), "track before send gas")
+			childCtx := ctx.WithGasMeter(sdk.NewGasMeter(types.TrackBeforeSendGasLimit))
```
Comment (AlpinYukseloglu) [inline x/tokenfactory/keeper/before_send.go:136]: Hmm a 100k gas limit seems very low here -- even the example blacklist contract costs 150k for a send right?

Comment (mattverse) [reply x/tokenfactory/keeper/before_send.go:136]: Mostly were random numbers for gas metering, do you have idea on how much it should be capped at?

Comment (ValarDragon) [reply x/tokenfactory/keeper/before_send.go:136]: def agreed 100k sounds too low, probably should be like 500k for now

Comment (mattverse) [reply x/tokenfactory/keeper/before_send.go]: Changed to 500k!

File: x/tokenfactory/keeper/before_send.go [hunk 2]
```diff
@@ -133,22 +133,14 @@ func (k Keeper) callBeforeSendListener(ctx sdk.Context, from, to sdk.AccAddress,
...
-				if err != nil {
-					return errorsmod.Wrapf(err, "failed to call before send hook for denom %s", coin.Denom)
-				}
-
-				// consume gas used for calling contract to the parent ctx
-				ctx.GasMeter().ConsumeGas(childCtx.GasMeter().GasConsumed(), "track before send gas")
+			childCtx := ctx.WithGasMeter(sdk.NewGasMeter(types.BeforeSendHookGasLimit))
+			_, err = k.contractKeeper.Sudo(childCtx.WithEventManager(em), cwAddr, msgBz)
+			if err != nil {
+				return errorsmod.Wrapf(err, "failed to call before send hook for denom %s", coin.Denom)
```
Comment (AlpinYukseloglu) [inline x/tokenfactory/keeper/before_send.go:139]: nit: would be good to have custom error type here

**Review by AlpinYukseloglu — APPROVED** [review]
This looks good as a patch, although with this change the distinction between `BlockBeforeSend` and `TrackBeforeSend` becomes obsolete and could be remerged into just `BeforeSend`.

Created an issue #6775 to track this – it's not necessarily urgent, but better to do it sooner since the API break will be more costly as more contracts onboard to using this feature

**Review by PaddyMc — APPROVED** [review]
LGTM

**Pull request merged**